setup.cfg

```diff
@@ -19,3 +19,40 @@ exclude = jedi/third_party/* .tox/*
 
 [pycodestyle]
 max-line-length = 100
+
+
+[mypy]
+# Ensure generics are explicit about what they are (e.g: `List[str]` rather than
+# just `List`)
+disallow_any_generics = True
+
+disallow_subclassing_any = True
+
+# Ensure that optional types are explicit, aiding clarity and ensuring that
+# consumers have the choice to enable these if they want to.
+no_implicit_optional = True
+strict_optional = True
+
+# Avoid creating future gotchas emerging from bad typing
+warn_redundant_casts = True
+warn_unused_ignores = True
+warn_return_any = True
+warn_unused_configs = True
+
+warn_unreachable = True
+
+# Require values to be explicitly re-exported; this makes things easier for
+# Flake8 too and avoids accidentally importing thing from the "wrong" place
+# (which helps avoid circular imports)
+implicit_reexport = False
```

Comment:
Just as a bit of further explanation. `__all__` does absolutely nothing apart from one thing: Limiting `from [...] import *`. This means that everyone can still access everything. It looks like people cannot do that, but it's one of those ugly Python hacks (like name mangling with double underscores).

You are obviously right that this would solve a few issues with flake8 as well, but I really despise `__all__`.

Reply:
Would you be up for marking specific files as `implicit_reexport = True` (we could also have Flake8 ignore F401 specifically in the same files), but keeping this default?
That would avoid `__all__` (which I agree isn't great), while still getting most of the strictness benefits.

Reply:
That would be fine for me (at least for now, we can still change it later).

Reply:
cefc363f64a32f63498bc594d5154f0ca3be09bf, which also changes to be strict about `F401` everywhere else.